### ISSUE TITLE
Parse data even if responses contain errors

### ISSUE THREAD

**Issue by mike-schmid** [graphql-java-generator/graphql-maven-plugin-project#195]
Hi
I use the graphql-maven-plugin in client mode (with spring boot).

Is there a way to get the data of a response that also contains errors?
Currently there is [this check](https://github.com/graphql-java-generator/graphql-maven-plugin-project/blob/d3b5651f1dfdb4e52336a66db477f70538ba7e28/graphql-java-client-runtime/src/main/java/com/graphql_java_generator/client/request/AbstractGraphQLRequest.java#L531) in place that throws an exception if a response contains errors:

```
// Does this response contain errors?
if (response.getErrors() != null && response.getErrors().size() > 0) {
		throw new GraphQLRequestExecutionException(response.getErrors());
}

// No error, let's parse the response data
try {
...
}
```

In my case, the errors do not really affect the transmitted data and I want that data to be parsed and returned anyway.
is there a way to get around this check? Changing the API is not possible.

**Comment by mike-schmid** [comment]
Any suggestions for a workaround?

**Comment by etienne-sf** [comment]
Hum, hum,

I don't see an easy workaround. 

I see two possible corrections:
1) Attach the parsed data to the ´ GraphQLRequestExecutionException´, so that read it. 
2) Find a way to define whether the exception should be thrown or not. But this would probably complex, and would break existing client (which is bad). 

Would solution 1 be ok for your use case ?

**Comment by mike-schmid** [comment]
Thanks for the reply.
Solution 1 is okay for our use case. But because with the current implementation in the [AbstractGraphQLRequest.java](https://github.com/graphql-java-generator/graphql-maven-plugin-project/blob/d3b5651f1dfdb4e52336a66db477f70538ba7e28/graphql-java-client-runtime/src/main/java/com/graphql_java_generator/client/request/AbstractGraphQLRequest.java#L531) it might be hard to overwrite this functionality ourselves (without forking the whole project).
Would you mind adding this change to a future release?
It might be useful for other devs as well because having errors and actual data is a valid scenario in graphql.
(see https://spec.graphql.org/October2021/#example-072c4)

**Comment by etienne-sf** [comment]
Hello,

The correction for this issue has been released the 2.3 version.
I'll see to add it into the 1.x version

**Comment by jaggaer-sbudkevi** [comment]
Faced the same issue and is trying to override it somehow. My concern is the following (I'm not a PRO in graphql):
- any graphql returns 200 always if there are errors or not
- if errors exist they can be found next to the data field and easily be extracted.
That's why I was pretty surprised to found that generated code throws exceptions instead of just convert it to something like `DataFetcherResult.java`

**Comment by etienne-sf** [comment]
Hello,

There are two points here:

1. The plugin should parse the data, whether or not there are errors in the response
2. The behavior of the request when an error occurs

For point 1, it's a mistake. And I was pretty sure that it had been treated before releasing 2.3. But unfortunately, it was not. I'll change that.
For point 2, there is a discussion here.

My opinion is that when the error part is filled, it generally means that something unexpected happens. And the correct behavior, once in the java worlds to manage what's unexpected is to manage this in exceptions, so that you can handle exception at different levels of the code.
This still allows to manage the exception locally, to apply the needed behavior.

Etienne

**Comment by jaggaer-sbudkevi** [comment]
My concern is based on the current behaviour of com.graphql-java. It allows you to receive both errors and data. And before we migrated to your plugin, that's behaviour was source of truth for us. I doubt, that this java graphql is A MUST behaviour, just wanted to clarify my vision of the problem.

**Comment by etienne-sf** [comment]
> My concern is based on the current behaviour of com.graphql-java

Interesting. But not that simple: since the beginning of the plugin, GraphQL errors are translated into exceptions. Changing that would break the implementations based on the plugin.
I can't do that.
An alternative is to add a plugin parameter, that would control this part of the code generation. But there already are a lot of parameters. I'm not sure that adding complexity to the plugin is a good idea.
And the idea of the exception is to:
* Make the code simple, so that it manages only the standard behaviour (no error)
* Exception should be managed elsewhere (in a catch statement). I think that writing tests about the error tag after each GraphQL call is a bad practice. And actually, it would be exactly the same as adding a catch (but with another syntax).

**About the 200 code** : this a rather standard behaviour, when calling an external service (SOAP, GraphQL...). On this points REST is another beast, where the http status is the error code.
So when the exchange is ok on a technical level, the client receives a 200 code. This doesn't mean that the request is even valid. It just means that a valid json was sent. If this valid json contains an invalid GraphQL request, then the response is a 200 status, but with error in there (for instance, when you query a non-existing query or mutation, or a non-existing field).

The other point here is when **sending several queries/mutations** at once.
This is actually a good shower for this requirement. :)



**(temporary) conclusion:**
The plugin should:
* Keep throwing an exception when the _error_ tag is not empty (to not break the existing code, based on the plugin).
    * Adding a catch is functionally the same as checking if the _error_ tag is empty or not.
    * It adds this advantage that the test can be made in the method that executes the GraphQL call, or any of its caller (which is basically the idea behind the _exception_ stuff)
* Parse the _data_ tag before checking if there is an error.
    *  Then, if an exception is thrown, this exception would contain the response object (with the _data_ and _error_). And the _data_ should be mapped to the POJO generated by the plugin, from the GraphQL schema.

**Comment by etienne-sf** [comment]
Released in the 2.3.1 release
